hack/shoot-comparator/pkg/shoot/matcher.go

```diff
@@ -219,3 +183,208 @@ func (m *Matcher) NegatedFailureMessage(_ interface{}) string {
 func (m *Matcher) FailureMessage(_ interface{}) string {
 	return strings.Join(m.fails, "\n")
 }
+
+func mapMatchers(m map[string]string) []types.GomegaMatcher {
+	mLen := len(m)
+	if mLen == 0 {
+		return []types.GomegaMatcher{
+			gomega.BeEmpty(),
+		}
+	}
+
+	out := make([]types.GomegaMatcher, mLen)
+	index := 0
+	for key, val := range m {
+		matcher := gomega.HaveKeyWithValue(key, val)
+		out[index] = matcher
+		index++
+	}
+	return out
+}
+
+func val(v interface{}) string {
+	if reflect.ValueOf(v).Kind() == reflect.Pointer && reflect.ValueOf(v).IsNil() {
+		return ""
+	}
+
+	if reflect.ValueOf(v).Kind() == reflect.Pointer {
+		return fmt.Sprintf("%v", reflect.ValueOf(v).Elem())
+	}
+
+	return fmt.Sprintf("%v", v)
+}
+
+func idToleration(v interface{}) string {
+	toleration, ok := v.(v1beta1.Toleration)
+	if !ok {
+		panic("invalid type")
+	}
+	return fmt.Sprintf("%s:%s", toleration.Key, val(toleration.Value))
+}
+
+func tolerations(ts []v1beta1.Toleration) gstruct.Elements {
+	out := map[string]types.GomegaMatcher{}
+	for _, t := range ts {
+		ID := idToleration(t)
+		out[ID] = gstruct.MatchAllFields(gstruct.Fields{
+			"Key":   gomega.BeComparableTo(t.Key),
+			"Value": gomega.BeComparableTo(t.Value),
+		})
+	}
+	return out
+}
+
+func idProvider(v interface{}) string {
+	provider, ok := v.(v1beta1.DNSProvider)
+	if !ok {
+		panic("invalid type")
+	}
+
+	return fmt.Sprintf("%s:%s:%s",
+		val(provider.Type),
+		val(provider.SecretName),
+		val(provider.Primary))
+}
+
+func providers(ps []v1beta1.DNSProvider) gstruct.Elements {
+	out := map[string]types.GomegaMatcher{}
+	for _, p := range ps {
+		ID := idProvider(p)
+		domainsMatcher := gomega.BeNil()
+		if p.Domains != nil {
+			domainsMatcher = gstruct.PointTo(
+				gstruct.MatchFields(
+					gstruct.IgnoreMissing,
+					gstruct.Fields{
+						"Include": gomega.ContainElements(p.Domains.Include),
+					}))
+		}
+
+		out[ID] = gstruct.MatchFields(
+			gstruct.IgnoreMissing,
+			gstruct.Fields{
+				"Primary":    gomega.Equal(p.Primary),
+				"SecretName": gomega.Equal(p.SecretName),
+				"Type":       gomega.Equal(p.Type),
+				"Domains":    domainsMatcher,
+				"Zones":      gstruct.Ignore(),
+			})
+	}
+
+	return out
+}
+
+func newDNSMatcher(dns *v1beta1.DNS) types.GomegaMatcher {
+	if dns == nil {
+		return gomega.BeNil()
+	}
+
+	return gstruct.PointTo(gstruct.MatchFields(
+		gstruct.IgnoreMissing,
+		gstruct.Fields{
+			"Domain": gomega.BeComparableTo(dns.Domain),
+			"Providers": gstruct.MatchElements(
+				idProvider,
+				gstruct.IgnoreMissing,
+				providers(dns.Providers)),
+		}))
+}
+
+func newMaintenanceMatcher(spec v1beta1.ShootSpec) types.GomegaMatcher {
+	if spec.Maintenance == nil {
+		return gomega.BeNil()
+	}
+
+	return gstruct.PointTo(gstruct.MatchFields(gstruct.IgnoreExtras, gstruct.Fields{
```

Comment:
Shouldn't we explicitly ignore `timeWindow` and `confineSpecUpdateRollout` fields as we do for `spec.kubernetes`?